Expand .NET Standard facades during design-time builds when referencing a .NET Standard project that hasn't been built yet

dsplaisted opened the pull request:
Fixes #1403

I would like to fix #1393 together with this, which is a similar bug for .NET Standard 1.x projects.  However, I'm still assessing the risk of doing so by changing the TargetPlatformIdentifier from `Windows` to `Portable` for .NET Standard projects.

dsplaisted committed:
Expand .NET Standard facades during design-time builds when referencing a .NET Standard project that hasn't been built yet.

Fixes #1403

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets:
```diff
@@ -472,6 +472,34 @@ Copyright (c) .NET Foundation. All rights reserved.
+        BeforeTargets="GetTargetPath"
+        DependsOnTargets="GetTargetPathWithTargetPlatformMoniker">
+
+    <PropertyGroup>
```
ItemGroup?

rainersigwald commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets:
```diff
@@ -472,6 +472,34 @@ Copyright (c) .NET Foundation. All rights reserved.
+  
+  <Target Name="InjectTargetPathMetadata"
+        BeforeTargets="GetTargetPath"
+        DependsOnTargets="GetTargetPathWithTargetPlatformMoniker">
```
This will return different metadata when the project entry point is `Build` versus `GetTargetPath`, won't it? If building only `GetTargetPath`, the is-any-dependency-NETStandard target won't fire and the return value won't be transitive. But when returned from `Build`, `ResolveReferences` will have run and the return value will be transitive.

The behaviors could be aligned by adding reference resolution to this target's dependencies, but that would be a pretty big change; IDE and other specialized builds assume that `GetTargetPath` is very cheap and it wouldn't be if it was recursive.

rainersigwald replied:
This will return different metadata when the project entry point is `Build` versus `GetTargetPath`, won't it? If building only `GetTargetPath`, the is-any-dependency-NETStandard target won't fire and the return value won't be transitive. But when returned from `Build`, `ResolveReferences` will have run and the return value will be transitive.

The behaviors could be aligned by adding reference resolution to this target's dependencies, but that would be a pretty big change; IDE and other specialized builds assume that `GetTargetPath` is very cheap and it wouldn't be if it was recursive.

rainersigwald commented on src/Tasks/Microsoft.NET.Build.Extensions.Tasks/msbuildExtensions/Microsoft/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.NETFramework.targets:
```diff
@@ -35,6 +35,16 @@ Copyright (c) .NET Foundation. All rights reserved.
+         More context: https://github.com/dotnet/sdk/issues/1403
+         -->
+    <PropertyGroup Condition="'$(DependsOnNETStandard)' == '' AND '$(NETStandardInbox)' != 'true'">
+      <DependsOnNETStandard Condition="('%(_ResolvedProjectReferencePaths.TargetFrameworkIdentifier)' == '.NETStandard') And ('%(_ResolvedProjectReferencePaths.TargetFrameworkVersion)' >= '1.5')">true</DependsOnNETStandard>
```
Adding this check (with inverted version-number checking) to `ImplicitlyExpandDesignTimeFacades` (for System.Runtime facades) would fix #1393, right? That might be less invasive than returning `Portable`.

nguerrera replied:
That makes sense, but if we change common targets for that, then I think the TFI/TFV should be added there too.

dsplaisted committed:
Fix TargetPathWithTargetPlatformMoniker item injection

livarcocc commented:
@dsplaisted please, fill out the ASK mode template like https://github.com/dotnet/sdk/pull/1400 and tag Sri.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets:
```diff
@@ -472,6 +472,34 @@ Copyright (c) .NET Foundation. All rights reserved.
+    <ItemGroup>
+      <TargetPathWithTargetPlatformMoniker>
+        <TargetFrameworkIdentifier>$(TargetFrameworkIdentifier)</TargetFrameworkIdentifier>
+        <TargetFrameworkVersion>$(_TargetFrameworkVersionWithoutV)</TargetFrameworkVersion>
```
Is there a TargetPlatformVersion already? Does it have a v?

dsplaisted replied:
There is a `TargetPlatformVersion` and unlike `TargetFrameworkVersion`, it [does not start with a "v"](https://github.com/Microsoft/msbuild/blob/68bf979f5ea8f842d26d9ce48580383edfa3e944/src/Tasks/Microsoft.Common.CurrentVersion.targets#L87).

rainersigwald reviewed (approved):
This looks good. I think I'd _slightly_ prefer an `ItemDefinitionGroup` instead, but that's based on a very fuzzy "it oughta be faster" and you can safely ignore it.

nguerrera commented:
Can we add a test that simulates a design-time build and checks that the facades are resolved?

dsplaisted committed:
Use ItemDefinitionGroup to inject TargetPathWithTargetPlatformMoniker metadata

This avoids a circular target dependency on versions of MSBuild that don't have the refactored GetTargetPath (ie, our CI machines)

dsplaisted commented:
> I would like to fix #1393 together with this, which is a similar bug for .NET Standard 1.x projects. However, I'm still assessing the risk of doing so by changing the TargetPlatformIdentifier from Windows to Portable for .NET Standard projects.

After investigation, I think for now we should fix #1393 in a less risky way, which I've sent as a PR to MSBuild: https://github.com/Microsoft/msbuild/pull/2304

dsplaisted commented:
@srivatsn for approval

**Customer scenario**

Intellisense in project to project references from a .NET Framework 4.6.1 or higher project to a .NET Standard 2.0 project.

**Bugs this fixes:** 

#1403

**Workarounds, if any**

Build the solution, and then either make a change to the project that triggers a design-time build (such as adding a reference), or close the solution, delete the .vs folder, and then reopen.

**Risk**

Low.

**Performance impact**

No.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

The check to determine if the .NET Standard facades should be injected was depending on reading the metadata of the referenced assemblies.  For design-time builds, there may be no assembly on disk to read, so there needs to be an alternate way to detect that the referenced project targets .NET Standard.

**How was the bug found?**

Dogfooding

srivatsn commented:
Since this needs a vs insertion, can you create a VSO bug and paste in this template there?

nguerrera commented:
Livar created a vso bug already.

nguerrera commented:
[VSO 460283](https://devdiv.visualstudio.com/DevDiv/_workitems?id=460283&_a=edit&triage=true)

nguerrera commented:
[VSO 463935](https://devdiv.visualstudio.com/DevDiv/_workitems?newQuery=true&_a=edit&id=463935&triage=true)

dsplaisted committed:
Add test case that design item builds detect P2P references to .NET Standard projects

nguerrera commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExeWtihNetStandardLib.cs:
```diff
@@ -137,6 +137,65 @@ public void It_includes_netstandard(bool isSdk, ReferenceScenario scenario)
+
+            //  Verify that netstandard.dll was passed to compiler
+            var references = getCommandLineCommand.GetValues()
+                                .Where(arg => arg.StartsWith("/reference:"))
```
nit: indenting here for fluent calls is different than elswewhere in same file.

nguerrera commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExeWtihNetStandardLib.cs:
```diff
@@ -137,6 +137,65 @@ public void It_includes_netstandard(bool isSdk, ReferenceScenario scenario)
+                                .Where(arg => arg.StartsWith("/reference:"))
+                                .Select(arg => arg.Substring("/reference:".Length))
+                                //  Strip quotes if there are any
+                                .Select(r =>
```
I'd be tempted to say `r => r.Trim('"')` is good enough.

nguerrera commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExeWtihNetStandardLib.cs:
```diff
@@ -137,6 +137,65 @@ public void It_includes_netstandard(bool isSdk, ReferenceScenario scenario)
+                                        return r;
+                                    }
+                                })
+                                .ToList();
```
why ToList()?

dsplaisted replied:
I usually call ToList at the end of a chain of LINQ calls to avoid accidentally having it evaluate everything lazily multiple times, and because it makes debugging slightly easier.

nguerrera commented:
Received mail that this was approved.

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets:
```diff
@@ -472,6 +472,27 @@ Copyright (c) .NET Foundation. All rights reserved.
 
   <!--
   ============================================================
+                                      InjectTargetPathMetadata
```
This is no longer a target but you document it as if it were.
